examples/webgl_loader_texture_ktx2.html

```diff
@@ -150,10 +167,23 @@
 
 					const texture = await loader.loadAsync( `./textures/compressed/${path}` );
 					texture.minFilter = THREE.NearestMipmapNearestFilter;
+					texture.needsUpdate = true;
+
+					if ( path.startsWith( 'cube' )  ) {
+
+						texture.mapping = THREE.CubeUVReflectionMapping;
```

Comment:
I think this should be `CubeReflectionMapping`.

Reply:
I also think that you should assign instances of `DataCubeTexture` to `envMap` instead of `map`. Otherwise the renderer tries to use a 2D sampler (instead of `samplerCube`) with your cube textures which results in the respective runtime error.

But even then the texture definition fails at this check:

https://github.com/mrdoob/three.js/blob/0e486b120a86ad6bcc01099e574629fab888cdd3/src/renderers/webgl/WebGLTextures.js#L1139

Reply:
Maybe you can setup `DataCubeTexture` similar to: https://threejs.org/examples/webgl_materials_cubemap_mipmaps

Reply:
> I also think that you should assign instances of DataCubeTexture to envMap instead of map...

Hm, I see! For testing purposes, I'm resolving this with a new `TextureHelper` class to eliminate any guesswork on which sampler is used.

> Maybe you can setup the instance of DataCubeTexture inside KTX2Loader similar to: https://threejs.org/examples/webgl_materials_cubemap_mipmaps

So the idea here seems to be that each item in the `mipmaps` array is a DataCubeTexture itself? I've pushed code giving this a try, but DataCubeTexture appears black, or if a valid CompressedCubeTexture was previously used, the DataCubeTexture will display that instead. :/

Reply:
BTW: What was the reason for adding `CompressedCubeTexture`? You could already setup a normal `CompressedTexture` with six images but it seems there was a use case this kind of texture definition did not work in context of `Scene.background` (#25909).

Reply:
Mostly, the error about `THREE.WebGLState: TypeError: Failed to execute 'texSubImage2D'...` in the OP, in which I can't get WebGLRenderer to interpret the given data as a cube texture. I'm finding it difficult to construct valid textures in these layouts, and I think having distinct classes with documented parameters might help.

I imagine we don't _technically need_ as many texture classes as we have already, and everything else could just be parameterization. And in theory I'm fine with fewer classes, but we'd need documentation of those parameters. I haven't been able to figure it out from reading the renderer code so far, other than that it's theoretically supported. 😰

The difference between DataTexture and CompressedTexture is pretty fuzzy to me, especially since CompressedTexture can use internal formats that aren't technically compressed, but neither name seems like it encompasses the purpose of the other.

Reply:
Um, if I had to choose I would distinct between a normal texture class and a cube texture class. We already do that with `Texture` and ` CubeTexture` and I think it is consistent to transfer this policy to `DataTexture`. It's just important to me that we deprecate the "old" approach at some point and force users towards `DataCubeTexture`. Same for `CompressedCubeTexture`.

> Mostly, the error about THREE.WebGLState: TypeError: Failed to execute 'texSubImage2D'... in the OP, in which I can't get WebGLRenderer to interpret the given data as a cube texture.

Whether a texture is interpreted as a normal texture or cube texture depends on the shader. `WebGLUniforms` decides based on the uniform type what `setTexture*()` function is called from `WebGLTextures`. The given texture must be compatible to the respective sampler.

> The difference between DataTexture and CompressedTexture is pretty fuzzy to me, especially since CompressedTexture can use internal formats that aren't technically compressed, but neither name seems like it encompasses the purpose of the other.

`CompressedTexture` was initially introduced to easier implement a code path in `WebGLTextures` for calling the `compressed*` WebGL texture commands. There would be of course other ways to make that distinction (like checking the texture format) but I think using a separate class for that is fine. 

It's true that `CompressedTexture` can be configured with plain `RGBAFormat`. However, there is at least one use case where this is in some sense "required". `DDSLoader` supports loading S3TC textures that hold uncompressed RGBA data. Since the loader is derived from `CompressedTextureLoader`, it has to return an instance of `CompressedTexture`. Hence, `CompressedTexture` does support `RGBAFormat` (the only exception next compressed formats). If we want to stop that, DDSLoader` either needs to return a data texture or we stop support for uncompressed S3TC data completely.

Reply:
> It's just important to me that we deprecate the "old" approach at some point and force users towards DataCubeTexture. Same for CompressedCubeTexture.

I see! Yes that is fine with me. I don't fully understand the old approach, but that's ok.

> Whether a texture is interpreted as a normal texture or cube texture depends on the shader.

I think one issue here, maybe, was `scene.background`. It could be a Texture or CubeTexture and we need to decide based on the texture, not just the shader. Perhaps NodeMaterial has similar cases?

It sounds like we're OK with adding these classes then, but just need to figure out how to ensure they work correctly?

Reply:
Continued in:

- https://github.com/mrdoob/three.js/pull/29926